superset/db_engine_specs.py

```diff
@@ -83,6 +85,24 @@
 }
 
 
+class TimestampExpression(ColumnClause):
+    def __init__(self, expr: str, col: ColumnClause, **kwargs):
+        """Sqlalchemy class that can be can be used to render native column elements
+        respeting engine-specific quoting rules as part of a string-based expression.
+
+        :param expr: Sql expression with '{col}' denoting the locations where the col
+        object will be rendered.
+        :param col: the target column
+        """
+        super().__init__(expr, **kwargs)
+        self.col = col
+
+
+@compiles(TimestampExpression)
+def compile_timegrain_expression(element: TimestampExpression, compiler, **kw):
+    return element.name.replace('{col}', compiler.process(element.col, **kw))
```

Comment:
What is "name" of a ColumnClause ? 

Do we need to pass this further via the compiler ? Is this supposed to return a string back or is it supposed to return some sqlalchemy struct, and so we might have to do:

```
compiler.process(element.name.replace('{col}', compiler.process(element.col, **kw)))
```

Reply:
`name` is the part containing the column name/expression; think `SELECT name FROM table`. In this case `name` will be a string containing a token `{col}` to indicate where the native element `col` should be compiled. The reason the element is kept in native form is because at construction time we want to build an engine agnostic select statement (`SqlaTable.get_sqla_query()` in `connectors/sqla/models.py`). By doing it this way we can get a different value for `name` depending on the dialect when we are finally compiling the whole statement (`SqlaTable.get_query_str_extended()` in `connectors/sqla/models.py`). Example:
- `name` = `TRUNC(CAST({col} as DATE), 'MI')` (string)
- `col` = `MyMixedCaseCol` (Any Core SQLAlchemy element that can be compiled, but in this example a `ColumnClause` whose `name` = `MyMixedCaseCol`)

When compiling this using different dialects we would get different results:
- MSSQL: `TRUNC(CAST([MyMixedCaseCol] as DATE), 'MI')`
- Oracle: `TRUNC(CAST("MyMixedCaseCol" as DATE), 'MI')`

And so forth. I'm actually working on adding a more generalized version of this to SQLAlchemy, but since it requires fairly big changes to the codebase I don't expect to see it ship in a released version before the end of this year. Hence the simple version above.